README.md
telescripter
============

a jQuery plugin to display text in telescript fashion..

## Usage

Drop jQuery and the jquery-telescripter scripts inside your page.

Target an element with come text content and transform it into a telescripter display :

```js
$(function() { // wait DOM Loaded

  $("#telescripter-demo").telescripter();
});

```
If the text content is not in your element, you can provide it as well :

```js
$(function() { // wait DOM Loaded

  $("#telescripter-demo").telescripter({
    source: [
      "Telescripter\nA jQuery plugin",  // first page of centent
      "Code source:\nhttp://github.com/zipang/telescripter" // second page
    ]
  });
});

```
You can tweak it as you want by providing even more options :
```js
$(function() { // wait DOM Loaded

  $("#telescriptrer-demo").telescripter({
    prompt: " #>",    // that'll be the default line prompt
    cursor: true,     // display a cursor
    charDelay: 100,   // display a new character each 1/10sec
    lineDelay: 1000,  // wait 1sec between each line
    pageDelay: 5000,  // wait 5secs between each page display
    autostart: true,  // Automatically start the telescripter animation
    autoloop: true,   // Automatically loop on the provided content
    charEvents: true  // Get an additional 'newChar' event 
  });
})

```

## API

Once instanciated, the telescripter instance can be retrieved from the target elements it has been applied on :

```js
var telescripter = $("#telescripter-demo").data("telescripter");
```

You can now control the telescripter by invoking its public methods :

* `start()`
  Starts on a fresh page.
* `stop()`
  Stops the currently running animation.
* `pauseResume()`
  Toggle between the rendering flag, thus stoping or resuming the current animation.
* `cls([prompt])`
  Clear the screen (stopping if necessary the current animation.
  Optionally displays a new prompt.
* `displayPage(pageNumber[, callback])`
  Stops the currently running animation and displays the specified page.
* `displayRandomPage([callback])`
  Stops the currently running animation and displays a new random page.

## Events

When the animation is playing, the telescripter warns you of its different phases by sending custom events to its container object.

* `newPage` sent every time a new page is drawn.
* `newLine` sent every time a new line has been added to the display.
* `newChar` sent (optionaly) every time a new char is appended to the display.



CHANGELOG.txt
#CHANGELOG
##v0.3

###v0.3.1 2013-12-15

  * Added char events (optional)
  * Fixed: cursor dissapearing on space 

###v0.3.0 2013-12-09

  * Added cursor option.

##v0.2

###v0.2.2	2013-12-07

  * Added .displayPage(n)
  * Added .pauseResume()

###v0.2.1	2013-11-28

  * Added .displayRandomPage()
  * Added Test & Demo Page

###v0.2

  * New custom events : 'newpage' and 'newline'
  * Passed the rendered element to <DIV> to support separated HTML styles on prompt and line blocks. (span.prompt)
  * Support for a micro langage allowing custom prompt definitions on each  lines.

##v0.1 - 2013-11-17

  * Text only : the telescripter pages are rendered into a single <PRE> tag.
  * Loops through a text content through : pages / lines / chars with individual pageDelay, lineDelay, charDelay options
  * requestAnimationFrame() usage allows the animation to stop when the browser window loose focus.
  * the defaut jQuery plugin invocation on an element extracts the text from the element content and converts him into a telescripter.


jquery-telescripter-0.1.js
/**
 * TeleScripter - a jQuery Plugin
 * @version 0.1
 * @date 2013-11-17
 * @author zipang
 * @url http://github.com/zipang/telescripter
 * @licence MIT
 */
(function(w, $) {

	var DEFAULTS = {
		autostart: true,
		autoloop: true,
		prompt: "> ",
		pageDelay: 2500, // time between pages display
		lineDelay: 500,  // time between paragraphs display
		charDelay: 150   // time between individuals characters printing
	};

	/**
	 * TelescriptMachine
	 * @constructor
	 */
	function TelescriptMachine(target, options) {
		this.$display = $(target);
		this.options  = options;

		var text = this.$display.text();

		if (text && !options.source) {
			options.source = text.split("\n\n");
		}

		if (options.autostart) {
			this.start();
		}
	};

	// Define the prototype methods
	$.extend(
		TelescriptMachine.prototype, {

			start : function() {

				if (this.rendering) this.stop().cls();

				var i = 0, machine = this,
					pages = machine.options.source,
					len = pages.length,
					printPages = function() {

						if (machine.rendering) {
							machine.printPage(pages[i % len], printPages);
							i++;
						}
					};

				machine.rendering = true;
				printPages();
				return machine;
			},

			cls : function() {
				this.$renderElt = $("<pre>");
				this.$display.empty().append(this.$renderElt);
				this.refreshScreen(this.options.prompt);
				return this;
			},
			append : function(txt) {
				this.refreshScreen(this.text + txt);
				return this;
			},
			newline : function() {
				this.append("\n" + this.options.prompt);
				return this;
			},
			refreshScreen : function(text) {
				this.text = text;
				this.$renderElt.text(text);
				return this;
			},

			printPage : function(message, eop) {
				this.cls();

				var i = 0, machine = this,
					lines = message.split("\n"),
					lineDelay = machine.options.lineDelay,
					len = lines.length,
					printLines = function() {

						if (machine.rendering) {
							if (i < len) {
								requestAnimationFrame(function() {
									if (i > 0) machine.newline();
									machine.rendering = setTimeout(function() {
										machine.printLine(lines[i++], printLines);
									}, lineDelay);
								});

							} else {
								// Wait for the page transition delay
								// before calling callback
								machine.rendering = setTimeout(eop, machine.options.pageDelay);
							}
						}
					};

				printLines();
				return machine;
			},

			printLine : function(line, eol) {

				var chars = line.split(""),
					machine = this,
					i = 0, len = chars.length,
					delay = machine.options.charDelay,
					printChars = function() {
						var prog = machine.rendering;
						requestAnimationFrame(function() {

							if (machine.rendering != prog) return;

							if (i < len) {
								machine.refreshScreen(machine.text + chars[i++]);
								machine.rendering = setTimeout(printChars, delay);
							} else {
								eol(); // end-of-line callback
							}
						});
					};

				printChars();
			},
			stop: function() {
				clearTimeout(this.rendering);
				this.rendering = false;
				return this;
			}

		}
	);

	$.fn.telescripter = function (options) {

		return $(this).each(function(i, elt) {
			// Initialize
			var opt = $.extend({}, DEFAULTS, options),
			    telescripter = new TelescriptMachine(elt, opt);

			$(elt).data("telescripter", telescripter)
				.on("start", function() {
					telescripter.start();
				})
				.on("stop", function() {
					telescripter.stop();
				});
		});
	};
})(this, jQuery);



jquery-telescripter-0.2.js
/**
 * TeleScripter - a jQuery Plugin
 * @version 0.2.1
 * @date 2013-11-28
 * @author zipang
 * @url http://github.com/zipang/telescripter
 * @licence MIT
 */
(function(w, $) {

	var DEFAULTS = {
		autostart: true,
		autoloop: true,
		prompt: "> ",
		pageDelay: 2500, // time between pages display
		lineDelay: 500,  // time between paragraphs display
		charDelay: 150   // time between individuals characters printing
	};

	// INTERNAL TOOLS

	/**
	 * Extract extended info from a line like prompt
	 * Example:
	 *   [$, red]I will begin with a red "$ " prompt
	 * @param line
	 * @return {Object} text and prompt of the line or false if no prompt info was detected
	 */
	function extractPrompt(line) {
		if (line[0] !== "[") return false; // no prompt info

		// that's a prompt specification
		var lineInfo = {},
			parts = line.split("]"),
			TEXT=0, COLOR=1, CLASS=2, // index of each parts
			prompt = parts[0].substr(1).split(",");

		lineInfo.prompt = [
			"<span ",
			prompt[COLOR] ? "style='color: " + prompt[COLOR] + ";' " : "",
			"class='prompt",
			prompt[CLASS] ? " " + prompt[CLASS] : "",
			"'>",
			prompt[TEXT],
			"</span>"
		].join("");
		lineInfo.text = parts[1];

		return lineInfo;
	}

	/**
	 * TelescriptMachine
	 * @constructor
	 */
	function TelescriptMachine(target, options) {
		this.$display = $(target);
		this.options  = options;

		if (!options.source) {
			options.source = this.$display.text();
		}

		if (options.autostart) {
			this.start();
		}
	};

	// Define the prototype methods
	$.extend(
		TelescriptMachine.prototype, {

			start : function() {

				var i = 0,
					machine = this.rendering ? this.stop() : this.init(),
					pages = machine.pages,
					len = machine.pages.length,
					printPages = function() {

						if (machine.rendering && i<len) {
							machine.printPage(pages[i], i, printPages);
							i++;
						} else {
							if (machine.rendering && machine.options.autoloop) {
								i = 0;
								requestAnimationFrame(printPages);
							}
						}
					};

				machine.rendering = true;
				printPages();
				return machine;
			},

			init : function() {
				var pages = this.options.source;

				if (typeof pages === "string") {
					pages = pages.split("\n\n"); // take each pages apart
				}

				this.pages = pages;
				return this;
			},

			cls : function(displayPrompt) {
				this.$renderElt = $("<div>");
				this.$display.empty().append(this.$renderElt);
				this.text = "";
				if (displayPrompt) this.refreshScreen(this.options.prompt);
				return this;
			},
			append : function(txt) {
				this.refreshScreen(this.text + txt);
				return this;
			},
			newline : function(prompt) {
				this.append("<br>" + (prompt || this.options.prompt));
				return this;
			},
			refreshScreen : function(text) {
				this.text = text;
				this.$renderElt.html(text);
				return this;
			},
			displayRandomPage : function(callback) {

				var machine = this.rendering ? this.stop() : this.init(),
					rnd = Math.floor(Math.random()*machine.pages.length);

				machine.rendering = true;
				machine.printPage(machine.pages[rnd], rnd, callback);
				return machine;
			},

			printPage : function(page, pageNumber, eop) {

				var i = 0, machine = this,
					lines = page.split("\n"),
					lineDelay = machine.options.lineDelay,
					len = lines.length,
					printLines = function() {

						if (machine.rendering) {
							if (i < len) {
								requestAnimationFrame(function() {
									var prompt = machine.options.prompt,
										newLine = lines[i],
										lineInfo  = extractPrompt(newLine);

									if (lineInfo) {
										newLine = lineInfo.text;
										prompt = lineInfo.prompt;
									}

									machine.newline(prompt);
									machine.rendering = setTimeout(function() {
										machine.printLine(newLine, i, printLines);
										i++
									}, lineDelay);
								});

							} else if (eop) {
								// Wait for the page transition delay
								// before calling callback
								machine.rendering = setTimeout(eop, machine.options.pageDelay);
							} else { // no callback
								machine.rendering = false;
							}
						}
					};

				machine.cls(); // clear screen
				machine.$renderElt.trigger("newpage", [pageNumber+1]);

				printLines();
				return machine;
			},

			printLine : function(line, lineNumber, eol) {

				var chars = line.split(""),
					machine = this,
					i = 0, len = chars.length,
					delay = machine.options.charDelay,
					printChars = function() {
						var prog = machine.rendering;
						requestAnimationFrame(function() {

							if (machine.rendering != prog) return;

							if (i < len) {
								machine.refreshScreen(machine.text + chars[i++]);
								machine.rendering = setTimeout(printChars, delay);
							} else {
								machine.$renderElt.trigger("newline", [lineNumber+1]); // trigger the custom 'newline' event with the line number (1-based)
								eol(); // end-of-line callback
							}
						});
					};

				printChars();
			},
			stop: function() {
				clearTimeout(this.rendering);
				this.rendering = false;
				return this;
			}

		}
	);

	$.fn.telescripter = function (options) {

		return $(this).each(function(i, elt) {
			// Initialize
			var opt = $.extend({}, DEFAULTS, options),
			    telescripter = new TelescriptMachine(elt, opt);

			$(elt).data("telescripter", telescripter)
				.on("start", function() {
					telescripter.start();
				})
				.on("stop", function() {
					telescripter.stop();
				});
		});
	};
})(this, jQuery);



jquery-telescripter-0.2.1.js
/**
 * TeleScripter - a jQuery Plugin
 * @version 0.2.1
 * @date 2013-11-28
 * @author zipang
 * @url http://github.com/zipang/telescripter
 * @licence MIT
 */
(function(w, $) {

	var DEFAULTS = {
		autostart: true,
		autoloop: true,
		prompt: "> ",
		pageDelay: 2500, // time between pages display
		lineDelay: 500,  // time between paragraphs display
		charDelay: 150   // time between individuals characters printing
	};

	// INTERNAL TOOLS

	/**
	 * Extract prompt definition from a line if there is one
	 * Example:
	 *   [$, red]I will begin with a red "$ " prompt
	 * Full syntax:
	 *   <[prompt, color, className]><line>
	 * @param line
	 * @return {Object} text and prompt of the line or false if no prompt info was detected
	 */
	function extractPrompt(line) {
		if (line[0] !== "[") return false; // no prompt info

		// that's a prompt specification
		var lineInfo = {},
			parts = line.split("]"),
			TEXT=0, COLOR=1, CLASS=2, // index of each parts
			prompt = parts[0].substr(1).split(",");

		lineInfo.prompt = [
			"<span ",
			prompt[COLOR] ? "style='color: " + prompt[COLOR] + ";' " : "",
			"class='prompt",
			prompt[CLASS] ? " " + prompt[CLASS] : "",
			"'>",
			prompt[TEXT],
			"</span>"
		].join("");
		lineInfo.text = parts[1];

		return lineInfo;
	}

	/**
	 * TelescriptMachine
	 * @constructor
	 */
	function TelescriptMachine(target, options) {
		this.$display = $(target);
		this.options  = options;

		if (!options.source) {
			options.source = this.$display.text();
		}

		if (options.autostart) {
			this.start();
		}
	};

	// Define the prototype methods
	$.extend(
		TelescriptMachine.prototype, {

			start : function() {

				var i = 0,
					machine = this.rendering ? this.stop() : this.init(),
					pages = machine.pages,
					len = machine.pages.length,
					printPages = function() {

						if (machine.rendering && i<len) {
							machine.printPage(pages[i], i, printPages);
							i++;
						} else {
							if (machine.rendering && machine.options.autoloop) {
								i = 0;
								requestAnimationFrame(printPages);
							}
						}
					};

				machine.rendering = true;
				printPages();
				return machine;
			},

			init : function() {
				var pages = this.options.source;

				if (typeof pages === "string") {
					pages = pages.split("\n\n"); // take each pages apart
				}

				this.pages = pages;
				return this;
			},

			cls : function(displayPrompt) {
				this.$renderElt = $("<div>");
				this.$display.empty().append(this.$renderElt);
				this.text = "";
				if (displayPrompt) this.refreshScreen(this.options.prompt);
				return this;
			},
			append : function(txt) {
				this.refreshScreen(this.text + txt);
				return this;
			},
			newline : function(prompt) {
				this.append("<br>" + (prompt || this.options.prompt));
				return this;
			},
			refreshScreen : function(text) {
				this.text = text;
				this.$renderElt.html(text);
				return this;
			},
			displayRandomPage : function(callback) {

				var machine = this.rendering ? this.stop() : this.init(),
					rnd = Math.floor(Math.random()*machine.pages.length);

				machine.rendering = true;
				machine.printPage(machine.pages[rnd], rnd, callback);
				return machine;
			},

			printPage : function(page, pageNumber, eop) {

				var i = 0, machine = this,
					lines = page.split("\n"),
					lineDelay = machine.options.lineDelay,
					len = lines.length,
					printLines = function() {

						if (machine.rendering) {
							if (i < len) {
								requestAnimationFrame(function() {
									var prompt = machine.options.prompt,
										newLine = lines[i],
										lineInfo  = extractPrompt(newLine);

									if (lineInfo) {
										newLine = lineInfo.text;
										prompt = lineInfo.prompt;
									}

									machine.newline(prompt);
									machine.rendering = setTimeout(function() {
										machine.printLine(newLine, i, printLines);
										i++
									}, lineDelay);
								});

							} else if (eop) {
								// Wait for the page transition delay
								// before calling callback
								machine.rendering = setTimeout(eop, machine.options.pageDelay);
							} else { // no callback
								machine.rendering = false;
							}
						}
					};

				machine.cls(); // clear screen
				machine.$renderElt.trigger("newpage", [pageNumber+1]);

				printLines();
				return machine;
			},

			printLine : function(line, lineNumber, eol) {

				var chars = line.split(""),
					machine = this,
					i = 0, len = chars.length,
					delay = machine.options.charDelay,
					printChars = function() {
						var prog = machine.rendering;
						requestAnimationFrame(function() {

							if (machine.rendering != prog) return;

							if (i < len) {
								machine.refreshScreen(machine.text + chars[i++]);
								machine.rendering = setTimeout(printChars, delay);
							} else {
								machine.$renderElt.trigger("newline", [lineNumber+1]); // trigger the custom 'newline' event with the line number (1-based)
								eol(); // end-of-line callback
							}
						});
					};

				printChars();
			},
			stop: function() {
				clearTimeout(this.rendering);
				this.rendering = false;
				return this;
			}

		}
	);

	$.fn.telescripter = function (options) {

		return $(this).each(function(i, elt) {
			// Initialize
			var opt = $.extend({}, DEFAULTS, options),
			    telescripter = new TelescriptMachine(elt, opt);

			$(elt).data("telescripter", telescripter)
				.on("start", function() {
					telescripter.start();
				})
				.on("stop", function() {
					telescripter.stop();
				});
		});
	};
})(this, jQuery);



jquery-telescripter-0.2.2.js
/**
 * TeleScripter - a jQuery Plugin
 * @version 0.2.2
 * @date 2013-12-07
 * @author zipang
 * @url http://github.com/zipang/telescripter
 * @licence MIT
 */
(function(w, $) {

	var DEFAULTS = {
		autostart: true,
		autoloop: true,
		prompt: "> ",
		pageDelay: 2500, // time between pages display
		lineDelay: 500,  // time between paragraphs display
		charDelay: 150   // time between individuals characters printing
	};

	// INTERNAL TOOLS

	/**
	 * Extract prompt definition from a line if there is one
	 * Example:
	 *   [$, red]I will begin with a red "$ " prompt
	 * Full syntax:
	 *   <[prompt, color, className]><line>
	 * @param line
	 * @return {Object} text and prompt of the line or false if no prompt info was detected
	 */
	function extractPrompt(line) {
		if (line[0] !== "[") return false; // no prompt info

		// that's a prompt specification
		var lineInfo = {},
			parts = line.split("]"),
			TEXT=0, COLOR=1, CLASS=2, // index of each parts
			prompt = parts[0].substr(1).split(",");

		lineInfo.prompt = [
			"<span ",
			prompt[COLOR] ? "style='color: " + prompt[COLOR] + ";' " : "",
			"class='prompt",
			prompt[CLASS] ? " " + prompt[CLASS] : "",
			"'>",
			prompt[TEXT],
			"</span>"
		].join("");
		lineInfo.text = parts[1];

		return lineInfo;
	}

	/**
	 * TelescriptMachine
	 * @constructor
	 */
	function TelescriptMachine(target, options) {
		this.$display = $(target);
		this.options  = options;

		if (!options.source) {
			options.source = this.$display.text();
		}

		if (options.autostart) {
			this.start();
		}
	};

	// Define the prototype methods
	$.extend(
		TelescriptMachine.prototype, {

			start : function() {

				var i = 0,
					machine = this.rendering ? this.stop() : this.init(),
					pages = machine.pages,
					len = machine.pages.length,
					printPages = function() {

						if (machine.rendering && i<len) {
							machine.printPage(pages[i], i, printPages);
							i++;
						} else {
							if (machine.rendering && machine.options.autoloop) {
								i = 0;
								requestAnimationFrame(printPages);
							}
						}
					};

				machine.rendering = true;
				printPages();
				return machine;
			},

			init : function() {
				var pages = this.options.source;

				if (typeof pages === "string") {
					pages = pages.split("\n\n"); // take each pages apart
				}

				this.pages = pages;
				return this;
			},

			cls : function(displayPrompt) {
				this.$renderElt = $("<div>");
				this.$display.empty().append(this.$renderElt);
				this.text = "";
				if (displayPrompt) this.refreshScreen(this.options.prompt);
				return this;
			},
			append : function(txt) {
				return this.refreshScreen(this.text + txt);
			},
			newline : function(prompt) {
				this.append("<br>" + (prompt || this.options.prompt));
				return this;
			},
			refreshScreen : function(text) {
				this.text = text;
				this.$renderElt.html(text);
				return this;
			},
			displayRandomPage : function(callback) {

				var rnd = Math.floor(Math.random()*this.pages.length);

				return this.displayPage(rnd, callback);
			},
			displayPage : function(pageNumber, callback) {

				var machine = this.rendering ? this.stop() : this.init();

				machine.rendering = true;
				machine.printPage(machine.pages[pageNumber], pageNumber, callback);
				return machine;
			},

			printPage : function(page, pageNumber, eop) {

				var i = 0, machine = this,
					lines = page.split("\n"),
					lineDelay = machine.options.lineDelay,
					len = lines.length,
					printLines = function() {

						if (machine.rendering) {
							if (i < len) {
								requestAnimationFrame(machine.resume = function() {
									var prompt = machine.options.prompt,
										newLine = lines[i],
										lineInfo  = extractPrompt(newLine);

									if (lineInfo) {
										newLine = lineInfo.text;
										prompt = lineInfo.prompt;
									}

									machine.newline(prompt);
									machine.rendering = setTimeout(function() {
										machine.printLine(newLine, i, printLines);
										i++
									}, lineDelay);
								});

							} else if (eop) {
								// Wait for the page transition delay
								// before calling callback
								machine.rendering = setTimeout(eop, machine.options.pageDelay);
							} else { // no callback
								machine.rendering = false;
							}
						}
					};

				machine.cls(); // clear screen
				machine.$renderElt.trigger("newpage", [pageNumber+1]);

				printLines();
				return machine;
			},

			printLine : function(line, lineNumber, eol) {

				var chars = line.split(""),
					machine = this,
					i = 0, len = chars.length,
					delay = machine.options.charDelay,
					printChars = function() {
						var prog = machine.rendering;
						requestAnimationFrame(machine.resume = function() {

							if (machine.rendering === false) return;

							if (i < len) {
								machine.refreshScreen(machine.text + chars[i++]);
								machine.rendering = setTimeout(printChars, delay);
							} else {
								machine.$renderElt.trigger("newline", [lineNumber+1]); // trigger the custom 'newline' event with the line number (1-based)
								eol(); // end-of-line callback
							}
						});
					};

				printChars();
			},

			stop: function() {
				clearTimeout(this.rendering);
				this.rendering = false;
				return this;
			},

			pauseResume: function() {
				var machine = this;
				if (machine.rendering) {
					return machine.stop();
				} else {
					machine.rendering = true;
					return machine.resume();
				}
			}

		}
	);

	$.fn.telescripter = function (options) {

		return $(this).each(function(i, elt) {
			// Initialize
			var opt = $.extend({}, DEFAULTS, options),
			    telescripter = new TelescriptMachine(elt, opt);

			$(elt).data("telescripter", telescripter)
				.on("start", function() {
					telescripter.start();
				})
				.on("stop", function() {
					telescripter.stop();
				});
		});
	};
})(this, jQuery);



jquery-telescripter-0.3.1.js
/**
 * TeleScripter - a jQuery Plugin
 * @version 0.3.1
 * @date 2013-12-15
 * @author zipang
 * @url http://github.com/zipang/telescripter
 * @licence MIT
 */
(function(w, $) {

	var DEFAULTS = {
		autostart: true,
		autoloop: true,
		cursor: true,
		prompt: "> ",
		pageDelay: 2500,  // time between pages display
		lineDelay: 500,   // time between paragraphs display
		charDelay: 150,   // time between individuals characters printing
		charEvents: false // send events on each char ?
	};

	// CONST
	var CURSOR_OPEN = "<span class=\"cursor\">",
		CURSOR_CLOSE = "</span>";

	// INTERNAL TOOLS

	/**
	 * Extract prompt definition from a line if there is one
	 * Example:
	 *   [$, red]I will begin with a red "$ " prompt
	 * Full syntax:
	 *   <[prompt, color, className]><line>
	 * @param line
	 * @return {Object} text and prompt of the line or false if no prompt info was detected
	 */
	function extractPrompt(line) {
		if (line[0] !== "[") return false; // no prompt info

		// that's a prompt specification
		var lineInfo = {},
			parts = line.split("]"),
			TEXT=0, COLOR=1, CLASS=2, // index of each parts
			prompt = parts[0].substr(1).split(",");

		lineInfo.prompt = [
			"<span ",
			prompt[COLOR] ? "style='color: " + prompt[COLOR] + ";' " : "",
			"class='prompt",
			prompt[CLASS] ? " " + prompt[CLASS] : "",
			"'>",
			prompt[TEXT],
			"</span>"
		].join("");
		lineInfo.text = parts[1];

		return lineInfo;
	}

	/**
	 * TelescriptMachine
	 * @constructor
	 */
	function TelescriptMachine(target, options) {
		this.$renderElt = $(target);
		this.options  = options;

		if (!options.source) {
			options.source = this.$renderElt.text();
		}

		if (options.autostart) {
			this.start();
		}
	};

	// Define the prototype methods
	$.extend(
		TelescriptMachine.prototype, {

			start : function() {

				var i = 0,
					machine = this.rendering ? this.stop() : this.init(),
					pages = machine.pages,
					len = machine.pages.length,
					printPages = function() {

						if (machine.rendering === false) return;

						if ( i < len ) {
							machine.printPage(pages[i], i, printPages);
							i++;
						} else if (machine.options.autoloop) {
							i = 0;
							requestAnimationFrame(printPages);
						}
					};

				machine.rendering = true;
				printPages();
				return machine;
			},

			init : function() {
				var pages = this.options.source;

				if (typeof pages === "string") {
					pages = pages.split("\n\n"); // take each pages apart
				}

				this.pages = pages;
				return this;
			},

			cls : function() {
				return this.prompt(this.text = "");
			},
			append : function(txt) {
				return this.refreshScreen(this.text + txt);
			},
			prompt : function(prompt, lf) {
				return this.append(
					(lf ? "<br>" : "") 
						+ (prompt ? 
							"<span class=\"prompt\">" + prompt + "</span>"
							: "")
				);
			},
			refreshScreen : function(text) {
				var machine = this,
					renderedText = machine.text = text,
					lastPos, lastChar;

				if (machine.options.cursor) { // wrap the last character inside a cursor
					lastChar = text[lastPos = text.length - 1] || " ";

					if (/^[ >]/.test(lastChar)) { // space or '>'
						renderedText += (CURSOR_OPEN + "&nbsp;" + CURSOR_CLOSE);
						
					} else { // render the last char inside the cursor
						renderedText = text.substr(0, lastPos) + CURSOR_OPEN + lastChar + CURSOR_CLOSE;							
					}					
				}

				machine.$renderElt.html(renderedText);
				return machine;
			},
			
			displayRandomPage : function(callback) {

				var rnd = Math.floor(Math.random()*this.pages.length);

				return this.displayPage(rnd, callback);
			},
			displayPage : function(pageNumber, callback) {

				var machine = this.rendering ? this.stop() : this.init();

				machine.rendering = true;
				machine.printPage(machine.pages[pageNumber], pageNumber, callback);
				return machine;
			},

			printPage : function(page, pageNumber, eop) {

				var i = 0, machine = this,
					lines = page.split("\n"),
					lineDelay = machine.options.lineDelay,
					len = lines.length,
					printLines = function() {

						if (machine.rendering) {
							if (i < len) {
								requestAnimationFrame(machine.resume = function() {
									var prompt = machine.options.prompt,
										newLine = lines[i],
										lineInfo  = extractPrompt(newLine);

									if (lineInfo) {
										newLine = lineInfo.text;
										prompt = lineInfo.prompt;
									}

									machine.prompt(prompt, i > 0);
									machine.rendering = setTimeout(function() {
										machine.printLine(newLine, i, printLines);
										i++
									}, lineDelay);
								});

							} else if (eop) {
								// Wait for the page transition delay
								// before calling callback
								machine.rendering = setTimeout(eop, machine.options.pageDelay);
							} else { // no callback
								machine.rendering = false;
							}
						}
					};

				machine.cls(); // clear screen
				machine.$renderElt.trigger("newpage", [pageNumber+1]);

				printLines();
				return machine;
			},

			printLine : function(line, lineNumber, eol) {

				var chars = line.split(""),
					machine = this,
					char, i = 0, len = chars.push(" "),
					$renderElt = machine.$renderElt,
					delay = machine.options.charDelay,
					nextChar = machine.options.charEvents 
						? function() { 
							$renderElt.trigger("newChar", char = chars[i++]);
							return char; 
						} : function() { return chars[i++]; },
					printChars = function() {
						requestAnimationFrame(machine.resume = function() {

							if (machine.rendering === false) return;

							if (i < len) {
								machine.refreshScreen(machine.text + nextChar());
								machine.rendering = setTimeout(printChars, delay);
							} else {
								$renderElt.trigger("newline", [lineNumber+1]); // trigger the custom 'newline' event with the line number (1-based)
								eol(); // end-of-line callback
							}
						});
					};

				printChars();
			},

			stop: function() {
				clearTimeout(this.rendering);
				this.rendering = false;
				return this;
			},

			pauseResume: function() {
				var machine = this;
				if (machine.rendering) {
					return machine.stop();
				} else {
					machine.rendering = true;
					return machine.resume();
				}
			}

		}
	);

	$.fn.telescripter = function (options) {

		return $(this).each(function(i, elt) {
			// Initialize
			var opt = $.extend({}, DEFAULTS, options),
			    telescripter = new TelescriptMachine(elt, opt);

			$(elt).data("telescripter", telescripter)
				.on("start", function() {
					telescripter.start();
				})
				.on("stop", function() {
					telescripter.stop();
				});
		});
	};
})(this, jQuery);

